app/src/main/java/com/group80/uoftinder/firebase/realtime/ucWriter.java

```diff
@@ -0,0 +1,63 @@
+package com.group80.uoftinder.firebase.realtime;
+
+import com.google.firebase.database.DatabaseReference;
+import com.google.firebase.database.FirebaseDatabase;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class ucWriter<T> {
+    /**
+     * A reference to the root of the database
+     */
+    protected final DatabaseReference databaseReference;
+    /**
+     * A list of listeners, actions triggered on successful / failed uploads
+     */
+    protected List<RealtimeDbWriteListener> listeners;
+
+    /**
+     * Constructor, creates a realtime database writer
+     */
+    public ucWriter() {
+        this.databaseReference = FirebaseDatabase.getInstance().getReference();
+        this.listeners = new ArrayList<>();
+    }
+
+    /**
+     * Writes the data to the given structural location in the realtime database
+     *
+     * @param structure the file structure of the upload
+     * @param data      the data to be uploaded
+     */
+    public abstract void write(String[] structure, T data);
+
+    /**
+     * Adds a listener to the listener list
+     *
+     * @param listener the listener to be added
+     */
+    public void addListener(RealtimeDbWriteListener listener) {
+        this.listeners.add(listener);
+    }
+
+    /**
+     * Notifies the listeners that a successful upload has been made
+     *
+     * @param unused a placeholder
+     */
+    protected void notifySuccess(Void unused) {
```

Comment:
Quick question - did you consider putting the ```notifySuccess``` and ```notifyFailure``` methods into an ```Observable``` interface, which ```ucWriter``` could implement? This would be an easy way to include the Observer design pattern into your code.

Reply:
Treat listeners as observers. I'm already following the Observer design pattern --though instead of having one `update()` I have two -- on success and on fail. 

In line 51, `listener.onWriteSuccess(unused)` is similar to the `update()` of the observer we see in lecture, but only triggered when the writing task succeeded. Same for line 61 which is only triggered on fail. 
`ucWriter` should not implement implement these interfaces -- it shouldn't define the behaviour when the writing task succeed / failed (again ucWriter is only a UseCase). Instead, the object controlling the ucWriter should defind these tasks.